### PR TITLE
fix: #381 overwriting existing user records on every login

### LINKED ISSUE #381: Create new user only if an existing user with the same OIDC `sub` does not already exist
part of #353 

PR #378 is creating a bug where a new user record is created every time new _and_ existing users login.

### PR DESCRIPTION
**Pull request by llorllale**
~~depends on #387~~ 

fixes #381 

Signed-off-by: George Aristy <george.aristy@securekey.com>

### CODE REVIEW

File: pkg/restapi/oidc/persistedData.go [hunk 1]
```diff
@@ -61,4 +60,13 @@ func (e *persistedData) put(k string, v interface{}) error {
 	return e.s.Put(k, bits)
 }
 
+func (e *persistedData) getEndUser(sub string) (*endUser, error) {
```
Comment (aholovko) [inline pkg/restapi/oidc/persistedData.go:63]: looks like that the returned `endUser` is not used, maybe we don't want to get the user but, instead, check if he exists - if this is the intention plz update the name of that function to reflect that

**Review by aholovko — APPROVED** [review]

File: pkg/restapi/oidc/persistedData.go (hunk 1 above)
Comment (llorllale) [reply pkg/restapi/oidc/persistedData.go:63]: The general intent is to populate the frontend with the user's name etc.

**Review by sudeshrshetty — APPROVED** [review]

File: pkg/restapi/oidc/operations.go [hunk 2]
```diff
@@ -168,16 +167,26 @@ func (o *Operation) oidcCallbackHandler(w http.ResponseWriter, r *http.Request)
...
+		err = newPersistedData(o.store.users).put(user.Sub, user)
+		if err != nil {
+			common.WriteErrorResponsef(w, logger,
+				http.StatusInternalServerError, "failed to persist user data: %s", err.Error())
+
+			return
+		}
+	}
+
+	err = newPersistedData(o.store.tokens).put(user.Sub, &endUserTokens{
```
Comment (Baha-sk) [inline pkg/restapi/oidc/operations.go:188]: returned `err` of put() call is not checked here, also why not use the same put() call for `err == nil` and `errors.Is(err, storage.ErrValueNotFound)` above?

Comment (llorllale) [reply pkg/restapi/oidc/operations.go:188]: @Baha-sk 

> returned `err` of put() call is not checked here

`err` is checked right below this line.

> also why not use the same put() call for `err == nil` and `errors.Is(err, storage.ErrValueNotFound)` above?

This statement is saving `endUserTokens`. The one above is saving `endUser`.

Comment (Baha-sk) [reply pkg/restapi/oidc/operations.go:188]: I see, it was in the unchanged code section that I haven't opened. Sounds good then.

**Review by Baha-sk — APPROVED** [review]

**Pull request merged**